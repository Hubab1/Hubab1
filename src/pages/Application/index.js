import React, { Component } from 'react';
import { Route, Switch } from 'react-router-dom';

import { ROUTES } from 'constants/constants';
import ApplicationCompletePage from './pages/ApplicationComplete';
import ApplicationApprovedPage from './pages/ApplicationApproved';
import ApplicationDeniedPage from './pages/ApplicationDenied';
import ApplicationCancelledPage from './pages/ApplicationCancelled';

class ApplicationContainer extends Component {
    render() {
        return (
            <Switch>
                <Route path={ROUTES.APP_COMPLETE} component={ApplicationCompletePage} />
                <Route path={ROUTES.APP_APPROVED} component={ApplicationApprovedPage} />
                <Route path={ROUTES.APP_DENIED} component={ApplicationDeniedPage} />
                <Route path={ROUTES.APP_CANCELLED} component={ApplicationCancelledPage} />
            </Switch>
        );
    }
}

export default ApplicationContainer;
