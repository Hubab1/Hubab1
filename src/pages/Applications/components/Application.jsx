import React, { useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { isEmpty } from 'lodash';
import clsx from 'clsx';
import {
    makeStyles,
    Card,
    CardHeader,
    CardContent,
    Collapse,
    IconButton,
    Typography,
    Divider,
} from '@material-ui/core';
import { ExpandMore as ExpandMoreIcone } from '@material-ui/icons';

import {
    APPLICATION_STATUSES_LABELS,
    APPLICANT_ROLE_LABELS,
    APPLICATION_STATUSES_COLORS,
    APPLICATION_STATUS_DENIED,
} from 'constants/constants';
import ActionButton from 'common-components/ActionButton/ActionButton';
import { withRouter } from 'react-router';
import { connect } from 'react-redux';
import { fetchRenterProfile, selectors } from 'reducers/renter-profile';

const useStyles = makeStyles((theme) => ({
    root: {
        marginBottom: theme.spacing(2),
        '&:last-child': {
            marginBottom: 0,
        },
    },
    expand: {
        transform: 'rotate(0deg)',
        marginLeft: 'auto',
        transition: theme.transitions.create('transform', {
            duration: theme.transitions.duration.shortest,
        }),
    },
    expandOpen: {
        transform: 'rotate(180deg)',
    },
    divider: {
        margin: theme.spacing(0, 2),
    },
    typography: {
        color: '#828796',
        marginBottom: theme.spacing(0.5),

        '& span': {
            color: 'black',
        },
    },
    title: {
        color: '#000',
    },
    applicationId: {
        color: '#828796',
        display: 'block',
        marginTop: theme.spacing(2),
    },
}));

export function Application({ application = {}, isActive = true, fetchRenterProfile, initialPage, history }) {
    const { id, status, lease_start_date, lease_term, fees_breakdown, role, unit, community } = application;
    const classes = useStyles();
    const [expanded, setExpanded] = useState(isActive);
    const title = unit?.unit_number ? `${community.display_name}, #${unit.unit_number}` : community.display_name;

    const handleExpandClick = useCallback(() => {
        setExpanded(!expanded);
    }, [expanded]);

    const [appSelected, setAppSelected] = useState(false);

    useEffect(() => {
        if (initialPage && appSelected) {
            setAppSelected(false);
            history.push(initialPage);
        }
    }, [initialPage, appSelected, history]);

    const handleApplicationClick = async (id) => {
        await fetchRenterProfile(id);
        setAppSelected(true);
    };

    if (isEmpty(application)) {
        return null;
    }

    const showOpenApplicationButton = isActive || application.status === APPLICATION_STATUS_DENIED;

    return (
        <Card className={classes.root} elevation={2}>
            <CardHeader
                data-testid="header"
                title={
                    <Typography className={clsx(classes.typography, classes.title)} variant="body1">
                        <b>{title}</b>
                    </Typography>
                }
                subheader={
                    <Typography className={classes.typography} variant="body1">
                        Application Status:{' '}
                        <span style={{ color: APPLICATION_STATUSES_COLORS[status] }}>
                            {APPLICATION_STATUSES_LABELS[status]}
                        </span>
                    </Typography>
                }
                action={
                    <IconButton
                        className={clsx(classes.expand, {
                            [classes.expandOpen]: expanded,
                        })}
                        onClick={handleExpandClick}
                    >
                        <ExpandMoreIcone />
                    </IconButton>
                }
            />
            <Collapse in={expanded}>
                <Divider className={classes.divider} />
                <CardContent data-testid="content">
                    <Typography className={classes.typography} variant="body1">
                        Move in Date:{' '}
                        <span>{lease_start_date ? moment(lease_start_date).format('MM/DD/YYYY') : '---'}</span>
                    </Typography>
                    <Typography className={classes.typography} variant="body1">
                        Lease Term: <span>{lease_term ? lease_term : '---'}</span>
                    </Typography>
                    <Typography className={classes.typography} variant="body1">
                        Monthly Rent:{' '}
                        <span>
                            {fees_breakdown?.monthly_fees?.total ? `$${fees_breakdown.monthly_fees.total}` : '---'}
                        </span>
                    </Typography>
                    <Typography className={classes.typography} variant="body1">
                        Role: <span>{`${APPLICANT_ROLE_LABELS[role]}`}</span>
                    </Typography>
                    <Typography className={classes.applicationId} variant="caption">
                        Application ID <span>{id}</span>
                    </Typography>
                    {showOpenApplicationButton && (
                        <ActionButton
                            variant="outlined"
                            marginTop={20}
                            marginBottom={10}
                            onClick={() => handleApplicationClick(id)}
                        >
                            Go To Application
                        </ActionButton>
                    )}
                </CardContent>
            </Collapse>
        </Card>
    );
}

Application.propTypes = {
    application: PropTypes.shape({
        id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
        status: PropTypes.number.isRequired,
        lease_start_date: PropTypes.string,
        lease_term: PropTypes.number,
        fees_breakdown: PropTypes.object,
        role: PropTypes.number.isRequired,
        unit: PropTypes.object,
        community: PropTypes.object.isRequired,
    }),
    isActive: PropTypes.bool,
    initialPage: PropTypes.string,
    communityId: PropTypes.string,
    fetchRenterProfile: PropTypes.func,
    history: PropTypes.object,
};

const mapStateToProps = (state) => ({
    communityId: state.siteConfig.basename,
    initialPage: selectors.selectDefaultInitialPage(state),
});

const mapDispatchToProps = {
    fetchRenterProfile,
};

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Application));
