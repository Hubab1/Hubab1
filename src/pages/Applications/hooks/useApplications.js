import { useCallback, useEffect, useState } from 'react';
import API from 'api/api';

import { useMayorLoader } from 'common-components/MayorLoader/MayorLoaderProvider';

export default function useApplications(errorMessage) {
    const { toggleLoader } = useMayorLoader();
    const [loading, setIsLoading] = useState(false);
    const [error, setError] = useState(null);
    const [applications, setApplications] = useState([]);

    const getApplications = useCallback(async () => {
        setIsLoading(true);

        try {
            const response = await API.getApplications();
            const applications =
                response?.map(({ application, role, last_activity }) => {
                    return {
                        ...application,
                        role,
                        lastActivity: last_activity,
                    };
                }) || [];

            setApplications(applications);
        } catch {
            setError(errorMessage);
        } finally {
            setIsLoading(false);
        }
    }, [errorMessage]);

    useEffect(() => {
        getApplications();
    }, [getApplications]);

    useEffect(() => {
        toggleLoader(loading);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [loading]);

    return {
        loading,
        error,
        applications,
    };
}
