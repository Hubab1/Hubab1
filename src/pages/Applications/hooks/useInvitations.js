import { useCallback, useEffect, useState } from 'react';
import API from 'api/api';

import { useMayorLoader } from 'common-components/MayorLoader/MayorLoaderProvider';

export default function useInvitations(errorMessage) {
    const { toggleLoader } = useMayorLoader();
    const [loading, setIsLoading] = useState(false);
    const [error, setError] = useState(null);
    const [invitations, setInvitations] = useState([]);

    const getInvitations = useCallback(async () => {
        setIsLoading(true);

        try {
            const invitations = await API.getInvitations();
            setInvitations(invitations);
        } catch {
            setError(errorMessage);
        } finally {
            setIsLoading(false);
        }
    }, [errorMessage]);

    useEffect(() => {
        getInvitations();
    }, [getInvitations]);

    useEffect(() => {
        toggleLoader(loading);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [loading]);

    return {
        loading,
        error,
        invitations,
    };
}
