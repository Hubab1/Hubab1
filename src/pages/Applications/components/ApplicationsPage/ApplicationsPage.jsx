import React, { useMemo } from 'react';
import { makeStyles } from '@material-ui/core';

import {
    ACTIVE_APPLICATION_STATUSES,
    PAST_APPLICATION_STATUSES,
} from 'app/constants';
import useApplicationRoles from 'hooks/useApplicationRoles';
import Page from 'components/common/Page/Page';
import Application from '../Application/Application';
import { H3 } from 'assets/styles';

export const ERROR_MESSAGE = 'Oops, we\'re having trouble obtaining your applications. Please try again later.';

const useStyles = makeStyles(() => ({
    root: {
        textAlign: 'left',
    },
    section: {
        marginBottom: 32,

        '& h3': {
            marginBottom: 16,
        },
    },
}));

// TODO: page will get populated in a follow up pr
export function ApplicationsPage() {
    const classes = useStyles();
    const { error, applicationRoles } = useApplicationRoles();
    const notification = error && {
        type: 'error',
        messages: ERROR_MESSAGE
    };

    console.log({ applicationRoles });

    // TODO: test empty states
    const [active, past] = useMemo(() => {
        return [
            applicationRoles,
            applicationRoles
        ];
    }, [applicationRoles]);

    return (
        <Page
            className={classes.root}
            title="My Applications"
            notification={notification}
        >
            <div className={classes.section}>
                <H3>Active Applications</H3>
                {active.map((applicationRole, i) => {
                    return (
                        <Application
                            key={i}
                            application={applicationRole.application}
                            role={applicationRole.role}
                            isActive
                        />
                    );
                })}
            </div>
            <div className={classes.section}>
                <H3>Past Applications</H3>
                {past.map((applicationRole, i) => {
                    return (
                        <Application
                            key={i}
                            application={applicationRole.application}
                            role={applicationRole.role}
                            isActive={false}
                        />
                    );
                })}
            </div>
        </Page>
    );
}

export default ApplicationsPage;
