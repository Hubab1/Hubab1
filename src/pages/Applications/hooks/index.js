import useApplications from './useApplications';
import useInvitations from './useInvitations';

export { useApplications, useInvitations };
