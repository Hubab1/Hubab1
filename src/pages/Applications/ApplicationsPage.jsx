import React, { useMemo } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { makeStyles, Typography } from '@material-ui/core';

import { ACTIVE_APPLICATION_STATUSES, APPLICANT_ROLE_VALUES, PAST_APPLICATION_STATUSES } from 'constants/constants';
import * as hooks from './hooks';

import Page from 'common-components/Page/Page';
import Application from './components/Application';

export const ERROR_MESSAGE = "Oops, we're having trouble obtaining your applications. Please try again later.";

const useStyles = makeStyles(() => ({
    root: {
        textAlign: 'left',
    },
    section: {
        marginBottom: 32,

        '& h3': {
            marginBottom: 16,
            fontSize: 18,
            color: '#454B57',
        },
        '& h4': {
            fontSize: 16,
            color: '#828796',
        },
    },
}));

export const getAlreadyHasActiveApplicationForCommuntiyAsPrimaryError = (
    applicant,
    applications,
    community,
    unit,
    accessedAppByInvitationOrWebsite
) => {
    if (!accessedAppByInvitationOrWebsite || !community || !unit) return;
    if (applications?.length === 0) return;

    const activeApplicationsForCommunityAsPrimary = applications.filter((a) => {
        return (
            a.primary_applicant.id === applicant.id &&
            a.role === APPLICANT_ROLE_VALUES.ROLE_PRIMARY_APPLICANT &&
            a.community.id === community.id && ACTIVE_APPLICATION_STATUSES.includes(a.status)
        );
    });

    if (activeApplicationsForCommunityAsPrimary?.length > 0) {
        return (
            <span>
                Oops, it looks like you already have an active application for {community.display_name}. Please continue
                your application for unit {unit.unit_number}, or call our office at {community.contact_phone} if you
                would like to start another application at {community.display_name}.
            </span>
        );
    }
};

export function ApplicationsPage({ applicant, community, unit, accessedAppByInvitationOrWebsite }) {
    const classes = useStyles();
    const { loading, error, applications } = hooks.useApplications(ERROR_MESSAGE);
    const alreadyHasActiveApplicationForCommuntiyAsPrimaryError = useMemo(() => {
        return getAlreadyHasActiveApplicationForCommuntiyAsPrimaryError(
            applicant,
            applications,
            community,
            unit,
            accessedAppByInvitationOrWebsite
        );
    }, [applicant, applications, community, unit, accessedAppByInvitationOrWebsite]);

    const notification = (error || alreadyHasActiveApplicationForCommuntiyAsPrimaryError) && {
        type: 'error',
        messages: error || alreadyHasActiveApplicationForCommuntiyAsPrimaryError,
    };

    const [active, past] = useMemo(() => {
        return [
            applications.filter((application) => ACTIVE_APPLICATION_STATUSES.includes(application.status)),
            applications.filter((application) => PAST_APPLICATION_STATUSES.includes(application.status)),
        ];
    }, [applications]);

    const [showActiveEmptyState, showPastEmptyState] = useMemo(() => {
        return [!error && active.length === 0, !error && past.length === 0];
    }, [error, active, past]);

    return (
        <Page className={classes.root} title="My Applications" notification={notification} loading={loading}>
            <div data-testid="active-applications" className={classes.section}>
                <Typography variant="h3">Active Applications</Typography>
                {active.map((application) => {
                    return <Application key={application.id} application={application} isActive />;
                })}
                {showActiveEmptyState && (
                    <Typography variant="h4">{`You don't have any active applications.`}</Typography>
                )}
            </div>
            <div data-testid="past-applications" className={classes.section}>
                <Typography variant="h3">Past Applications</Typography>
                {past.map((application) => {
                    return <Application key={application.id} application={application} isActive={false} />;
                })}
                {showPastEmptyState && <Typography variant="h4">{`You don't have any past applications.`}</Typography>}
            </div>
        </Page>
    );
}

ApplicationsPage.propTypes = {
    applicant: PropTypes.object.isRequired,
    community: PropTypes.object,
    unit: PropTypes.object,
    accessedAppByInvitationOrWebsite: PropTypes.bool,
};

const mapStateToProps = (state) => ({
    applicant: state.applicant,
    community: state.configuration?.community,
    unit: state.configuration?.unit,
    accessedAppByInvitationOrWebsite: Boolean(state.siteConfig.hash),
});

const mapActionsToProps = null;

export default connect(mapStateToProps, mapActionsToProps)(ApplicationsPage);
