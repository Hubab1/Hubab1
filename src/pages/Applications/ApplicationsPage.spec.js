import React from 'react';
import { shallow, render } from 'enzyme';
import moment from 'moment';

import { ACTIVE_APPLICATION_STATUSES, PAST_APPLICATION_STATUSES, APPLICANT_ROLE_VALUES } from 'constants/constants';
import Page from 'common-components/Page/Page';
import * as hooks from './hooks';
import { ApplicationsPage, getAlreadyHasActiveApplicationForCommuntiyAsPrimaryError } from './ApplicationsPage';
import Application from './components/Application';

// Mock hooks, issue: https://stackoverflow.com/questions/53162001/typeerror-during-jests-spyon-cannot-set-property-getrequest-of-object-which
jest.mock('./hooks', () => ({
    useApplications: jest.fn(),
}));

const mockUseApplications = (returnValue = {}) => {
    return jest.spyOn(hooks, 'useApplications').mockReturnValue({
        loading: false,
        error: undefined,
        applications: [],
        ...returnValue,
    });
};

describe('ApplicationsPage', () => {
    let props = {};
    beforeEach(() => {
        props = {
            applicant: { id: 1 },
            community: null,
            unit: null,
            accessedAppByInvitationOrWebsite: false,
        };
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    it('useApplications error shows error notification', () => {
        mockUseApplications({ error: 'Woops', loading: true });
        const wrapper = shallow(<ApplicationsPage {...props} />);

        expect(wrapper.html().includes('My Applications')).toBe(true);
        expect(wrapper.html().includes('Woops')).toBe(true);
        expect(wrapper.html().includes('Active Applications')).toBe(false);
        expect(wrapper.html().includes('Past Applications')).toBe(false);
    });

    it('useApplications loading shows empty page', () => {
        mockUseApplications({ loading: true });
        const wrapper = shallow(<ApplicationsPage {...props} />);

        expect(wrapper.html().includes('My Applications')).toBe(true);
        expect(wrapper.html().includes('Active Applications')).toBe(false);
        expect(wrapper.html().includes('Past Applications')).toBe(false);
    });

    it('useApplications success without application shows sections with empty copy', () => {
        mockUseApplications({ applications: [] });
        const wrapper = shallow(<ApplicationsPage {...props} />);

        expect(wrapper.html().includes('My Applications')).toBe(true);
        expect(wrapper.html().includes('Active Applications')).toBe(true);
        expect(wrapper.html().includes('You don&#x27;t have any active applications.')).toBe(true);
        expect(wrapper.html().includes('Past Applications')).toBe(true);
        expect(wrapper.html().includes('You don&#x27;t have any past applications.')).toBe(true);
    });

    it('useApplications success with application shows active and past application sections', () => {
        const applications = [
            {
                id: 'active-application',
                status: ACTIVE_APPLICATION_STATUSES[0],
                role: APPLICANT_ROLE_VALUES.ROLE_PRIMARY_APPLICANT,
                community: { display_name: 'community' },
                lastActivity: moment('2020-01-02').format('MM-DD-YYYY'),
            },
            {
                id: 'past-application',
                status: PAST_APPLICATION_STATUSES[0],
                role: APPLICANT_ROLE_VALUES.ROLE_CO_APPLICANT,
                community: { display_name: 'community' },
                lastActivity: moment('2020-01-01').format('MM-DD-YYYY'),
            },
        ];
        mockUseApplications({ applications });
        const wrapper = shallow(<ApplicationsPage {...props} />);
        const activeApplications = wrapper.find('[data-testid="active-applications"]').find(Application);
        const pastApplications = wrapper.find('[data-testid="past-applications"]').find(Application);

        expect(activeApplications.length).toBe(1);
        expect(pastApplications.length).toBe(1);
    });

    it('show already has active application for community as primary applicant', () => {
        props = {
            applicant: { id: 1 },
            community: { id: 1, display_name: 'Community Name', contact_phone: '(858) 485-9199' },
            unit: { id: 1, unit_number: '207' },
            accessedAppByInvitationOrWebsite: true,
        };

        const applications = [
            {
                id: 1,
                status: ACTIVE_APPLICATION_STATUSES[0],
                role: APPLICANT_ROLE_VALUES.ROLE_PRIMARY_APPLICANT,
                primary_applicant: props.applicant,
                community: props.community,
                lastActivity: moment('2020-01-02').format('MM-DD-YYYY'),
            },
        ];
        mockUseApplications({ applications });

        const wrapper = shallow(<ApplicationsPage {...props} />);
        const page = wrapper.find(Page);
        const renderedError = render(page.prop('notification').messages).html();

        expect(renderedError).toEqual(
            'Oops, it looks like you already have an active application for Community Name. ' +
            'Please continue your application for unit 207, or call our office at (858) 485-9199 ' +
            'if you would like to start another application at Community Name.'
        );
    });
});
