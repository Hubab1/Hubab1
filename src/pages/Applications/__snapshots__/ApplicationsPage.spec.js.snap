// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationsPage useApplications success with application shows active and past application sections 1`] = `
<Page
  className="makeStyles-root-1"
  loading={false}
  title="My Applications"
>
  <div
    className="makeStyles-section-2"
    data-testid="active-applications"
  >
    <ForwardRef(WithStyles)
      variant="h3"
    >
      Active Applications
    </ForwardRef(WithStyles)>
    <UNDEFINED
      application={
        Object {
          "community": Object {
            "display_name": "community",
          },
          "id": "active-application",
          "lastActivity": "01-02-2020",
          "role": 10,
          "status": 5,
        }
      }
      isActive={true}
    />
  </div>
  <div
    className="makeStyles-section-2"
    data-testid="past-applications"
  >
    <ForwardRef(WithStyles)
      variant="h3"
    >
      Past Applications
    </ForwardRef(WithStyles)>
    <UNDEFINED
      application={
        Object {
          "community": Object {
            "display_name": "community",
          },
          "id": "past-application",
          "lastActivity": "01-01-2020",
          "role": 20,
          "status": 25,
        }
      }
      isActive={false}
    />
  </div>
</Page>
`;

exports[`ApplicationsPage useInvitations success with invitations shows invitations 1`] = `
<Page
  className="makeStyles-root-1"
  loading={false}
  title="My Applications"
>
  <div
    className="makeStyles-section-2"
    data-testid="active-applications"
  >
    <ForwardRef(WithStyles)
      variant="h3"
    >
      Active Applications
    </ForwardRef(WithStyles)>
    <UNDEFINED
      application={
        Object {
          "community": Object {
            "display_name": "community",
          },
          "lastActivity": "01-02-2020",
          "status": 5,
        }
      }
      invitee={
        Object {
          "id": 123,
          "role": 20,
        }
      }
      isActive={true}
    />
    <UNDEFINED
      application={
        Object {
          "community": Object {
            "display_name": "community",
          },
          "lastActivity": "01-01-2020",
          "status": 5,
        }
      }
      invitee={
        Object {
          "id": 123,
          "role": 20,
        }
      }
      isActive={true}
    />
    <ForwardRef(WithStyles)
      variant="h4"
    >
      You don't have any active applications.
    </ForwardRef(WithStyles)>
  </div>
  <div
    className="makeStyles-section-2"
    data-testid="past-applications"
  >
    <ForwardRef(WithStyles)
      variant="h3"
    >
      Past Applications
    </ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      variant="h4"
    >
      You don't have any past applications.
    </ForwardRef(WithStyles)>
  </div>
</Page>
`;
